### PR TITLE
Use Array.Empty<T> for LINQ's Enumerable.Empty<TResult>

### PR DESCRIPTION
**Pull request by justinvp**

### CODE REVIEW

File: src/System.Linq/src/System/Linq/Enumerable.cs [hunk 1]
```diff
@@ -1452,7 +1452,10 @@ private static IEnumerable<TResult> RepeatIterator<TResult>(TResult element, int
 
         public static IEnumerable<TResult> Empty<TResult>()
         {
-            return EmptyEnumerable<TResult>.Instance;
+            // This is a more memory-intensive empty enumerable that allocates a new enumerator each time.
```
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: I don't believe this comment is accurate any more.  Enumerating an empty array should not allocate a new enumerator: https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Array.cs#L2717

Comment (justinvp) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: Excellent! I'll remove the comment.

**Comment by stephentoub** [comment]
cc: @VSadov

File: src/System.Linq/src/System/Linq/Enumerable.cs (hunk 1 above)
Comment (justinvp) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: As an aside, correct me if I am wrong, but it looks like `SZGenericArrayEnumerator<T>.Empty` is being greedily allocated as soon as `SZGenericArrayEnumerator<T>` is used for the first time (first time the generic `GetEnumerator` is called). Which means lots of possible unnecessary empty enumerator allocations for each T. Wouldn't it be better if this was lazily allocated only when an empty enumerator was actually needed? Something like `EmptySZGenericArrayEnumerator<T>.Instance`. (I can open an issue/PR on CoreCLR to continue the discussion there, if warranted).

**Comment by VSadov** [comment]
LGTM

File: src/System.Linq/src/System/Linq/Enumerable.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: Yeah, that's probably a good idea.

Comment (justinvp) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: Never mind -- I tested it out and the runtime is doing the right thing.

**Pull request merged**